### PR TITLE
Update third-parties on Orchest update

### PR DESCRIPTION
**Pull request by nhaghighat**
## Description

In this PR, the controller checks the deployed manifests of third parties with newly generated manifests and if they are not equal, it updates the helm releases. in this way even if we upgrade the helm templates of third parties, the controller still can update them.

Fixes: #1092 

## Checklist

- [x] I have manually tested my changes and I am happy with the result.
- [x] The documentation reflects the changes.
- [x] The PR branch is set up to merge into `dev` instead of `master`.

### CODE REVIEW

File: services/orchest-controller/pkg/addons/helm_deployer.go [hunk 1]
```diff
@@ -25,40 +25,57 @@ func NewHelmDeployer(name, deployDir string, valuesPath string) Addon {
...
+
+	// First we need to check if there is already a release
+	oldConfig, err := helm.GetReleaseConfig(ctx, releaseName, namespace)
+	if err == nil {
+		// check if update is required or not
+		newConfig, err := helm.RunCommand(ctx, deployArgs.WithTemplate().Build())
+		if err != nil {
+			// Failed to get new config, probably it is best to not update
+			return err
+		}
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/addons/helm_deployer.go:61]: Probably best to add some logging when an error happens here

File: services/orchest-controller/pkg/addons/helm_deployer.go [hunk 2]
```diff
@@ -25,40 +25,57 @@ func NewHelmDeployer(name, deployDir string, valuesPath string) Addon {
...
+		// Unfortunately, the value returned from helm get manifest has 1 extra byte,
+		// so in order to compare it with the new config, we have to remove the tail of it.
+		if newConfig == oldConfig[:len(oldConfig)-1] {
+			// There is no need for update, return without err
+			return nil
+		}
+	}
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/addons/helm_deployer.go:68]: Is the byte always the same? Can we do trimming in a safer manner?

We are comparing string to string, is it a yaml or a json or anything easily parseable? I was wondering if it makes sense to actually parse it and do a deep equality check at a higher level of abstraction

File: services/orchest-controller/pkg/addons/addons.go [hunk 3]
```diff
@@ -35,8 +35,8 @@ func NewDefaultAddonsConfig() AddonsConfig {
 }
 
 type Addon interface {
-	// Installs addon if the config is changed
-	Enable(ctx context.Context, namespace string, config *orchestv1alpha1.ApplicationConfig) error
+	// Enable the addon. preInstallHooks should be called, before enabling the addon
+	Enable(ctx context.Context, preInstallHooks []func() error, namespace string, config *orchestv1alpha1.ApplicationConfig) error
```
Comment (fruttasecca) [inline services/orchest-controller/pkg/addons/addons.go:39]: I don't have the full context, would it make sense to have the implementors of this interface both have `RunPreinstallHooks` (or something like that) and `Enable`? So that the order of calling is not a concern of the implementor

File: services/orchest-controller/pkg/addons/helm_deployer.go (hunk 1 above)
Comment (nhaghighat) [reply services/orchest-controller/pkg/addons/helm_deployer.go:62]: If it returns error, the log will be printed here:
https://github.com/orchest/orchest/pull/1126/files#diff-ea00a1f5097a1bc613e94d7d1f0b13f2c785eab68b530ab5df8ed91f9b0b1773L596

File: services/orchest-controller/pkg/addons/helm_deployer.go (hunk 2 above)
Comment (nhaghighat) [reply services/orchest-controller/pkg/addons/helm_deployer.go]: I trimmed the string and it resolved the issue.

File: services/orchest-controller/pkg/helm/builder.go [hunk 4]
```diff
@@ -18,17 +19,27 @@ func NewHelmArgBuilder() *HelmArgBuilder {
...
 func (builder *HelmArgBuilder) WithUpgradeInstall() *HelmArgBuilder {
-	builder.args = append(builder.args, "upgrade", "--install")
+	builder.command = []string{"upgrade", "--install"}
 	return builder
 }
 
 func (builder *HelmArgBuilder) WithUnInstall() *HelmArgBuilder {
-	builder.args = append(builder.args, "uninstall")
+	builder.command = []string{"uninstall", "--install"}
+	return builder
```
[the remark this answers is not in the document]
Comment (nhaghighat) [reply services/orchest-controller/pkg/helm/builder.go]: good catch

File: services/orchest-controller/pkg/addons/addons.go (hunk 3 above)
Comment (nhaghighat) [reply services/orchest-controller/pkg/addons/addons.go:39]: Well, I want the hooks to be called if necessary, and the deployer has the information if the install/update is needed or not

**Comment by yannickperrenet** [comment]
I just ran a test and it worked perfectly. What I did (for completeness):

1. Build Orchest on this branch with a version tag of `v2022.07.3`
2. Change `podLabels` in `argo-workflows/orchest-values.yaml` so I can check whether the argo controller indeed get's the newly added labels after the update
3. Follow [the docs](https://docs.orchest.io/en/stable/development/development_workflow.html#test-updating-orchest) to update to `v2022.07.4` with `--dev`
4. Check whether the update succeeded and whether the Argo controller has the newly added labels

And it did! The Argo controller got redeployed :smile_cat: 

Great job :fire:

File: services/orchest-controller/deploy/thirdparty/docker-registry/orchest-values.yaml [hunk 5]
```diff
@@ -41,7 +41,7 @@ storage: filesystem
 
 tlsSecretName: registry-tls-secret
 secrets:
-  haSharedSecret: ""
+  haSharedSecret: "haSharedSecret"
```
Comment (yannickperrenet) [inline services/orchest-controller/deploy/thirdparty/docker-registry/orchest-values.yaml:44]: Was this just for testing or why is this needed?

File: services/orchest-controller/pkg/addons/helm_deployer.go [hunk 6]
```diff
@@ -25,40 +26,56 @@ func NewHelmDeployer(name, deployDir string, valuesPath string) Addon {
...
+		if err != nil {
+			// Failed to get new config, probably it is best to not update
+			return err
+		}
+		// Unfortunately, the value returned from helm get manifest has 1 extra byte,
+		if strings.TrimSpace(newConfig) == strings.TrimSpace(oldConfig) {
+			// There is no need for update, return without err
+			return nil
+		}
+	}
```
Comment (yannickperrenet) [inline services/orchest-controller/pkg/addons/helm_deployer.go:68]: It seems to me that this is the "most important" addition of this PR. However, I don't completely understand how the `oldConfig` and `newConfig` are constructed. So my question is: How do we create the `oldConfig` and how do we create the `newConfig` in order to be able to compare the two to see whether there are changes?

Asking, because when updating Orchest we first update the `orchest-controller` and so the old container is no longer running.

File: services/orchest-controller/deploy/thirdparty/docker-registry/orchest-values.yaml (hunk 5 above)
Comment (nhaghighat) [reply services/orchest-controller/deploy/thirdparty/docker-registry/orchest-values.yaml:44]: This line deserves a comment in the code, if we don't provide this value, a new one will be randomly generated and every time we want to see if the docker registry is updated or not, since this value is different we think the docker-registry is updated.

Comment (yannickperrenet) [reply services/orchest-controller/deploy/thirdparty/docker-registry/orchest-values.yaml:44]: Ahaa makes a ton of sense. A comment explaining exactly that would indeed be a good idea :100:

File: services/orchest-controller/pkg/addons/helm_deployer.go (hunk 6 above)
Comment (nhaghighat) [reply services/orchest-controller/pkg/addons/helm_deployer.go:68]: `helm` stores the manifests of the release in a `secret` inside k8s, and uses that to retrieve the deployed manifests.
regarding the new config, `helm` has an option to give us the manifest it is going to deploy without deploying it.

Comment (yannickperrenet) [reply services/orchest-controller/pkg/addons/helm_deployer.go:68]: Great explanation, thank you! Can you find the appropriate place to add that as a comment? :)
